### PR TITLE
Oil Calculation Functionality

### PR DESCRIPTION
**Pull request by gbasile**
#### TL;DR

First PR for my NCU project. 
RxVespa is an utility application for Vespa owner and will grow in functionalities with the time. The goal of this project is to get familiar with RxSwift. 
Design is not covered at all in this PR
#### Problem

Here we implement a simple utility to calculate the amount of oil that need to be mixed in the gasoline, as vintage Vespa does not have an oil mixed as modern motorbikes. 

For example, if I refill my Vespa with 5 liter of gasoline, I need to put 100ml of oil. This is because my motorbike need a 2% oil mix. 
#### Solution

**OilMixCalculator** is a simple class that calculate the quantity of oil depending of the gasoline in liter. 
As the user can manipulate 2 slider in order to change the oil % and the quantity of gasoline, **GasCalculatorViewModel** is initialised with 2 observable that represent these value being emitted by the user. 
**GasCalculatorViewModel** expose 3 observables, one for each one of the label that needs to be updated: 
1. oilMixObservable (Used by the the result label: "23ml")
2. oilValueObservable (Used by the label that describe the value of the oil slider: "2%")
3. gasObservable (Used by the label that describe the value of the gas slider: 4L)
##### Test(s) added

Yes, Added tests for the OilMixCalculator and for the GasCalculatorViewModel
##### Screenshots

vespa-gas

### CODE REVIEW

File: VespaGAS/ViewModels/GasCalculatorViewModel.swift [hunk 1]
```diff
@@ -1,5 +1,36 @@
 import Foundation
+import RxSwift
 
-public struct GasCalculatorViewModel {
+struct GasCalculatorViewModel {
+    private let disposeBag = DisposeBag()
+    private let oilValueSubject = BehaviorSubject(value: "0%")
```
Comment (gbasile) [inline VespaGAS/ViewModels/GasCalculatorViewModel.swift]: The reason why I'm using a BehaviorSubject is that it can be initialised with an initial value. 
Is it a good idea?

File: VespaGASTests/ViewModels/GasCalculatorViewModelTests.swift [hunk 2]
```diff
@@ -1,9 +1,158 @@
...
+            gasObservable: Observable.empty(),
+            oilPercentageObservable: Observable.empty()
+        )
+
+        _ = viewModel.oilMixObservable.subscribeNext { (oilML) in
+            XCTAssertEqual("0ml", oilML)
+            expectation.fulfill()
+            }.addDisposableTo(disposeBag)
+
+        self.waitForExpectationsWithTimeout(0.1, handler: nil)
```
Comment (dominicfreeston) [inline VespaGASTests/ViewModels/GasCalculatorViewModelTests.swift]: Have you looked into `RxBlocking` and `RxTest`? Been meaning to do it for a while 😄

File: VespaGAS/ViewModels/GasCalculatorViewModel.swift (hunk 1 above)
Comment (dominicfreeston) [inline VespaGAS/ViewModels/GasCalculatorViewModel.swift]: Considering you're using it to map values from the oilPercentageObservable, wouldn't a simple map of the input observable with a `startWith` do the job?

So something like:

``` swift
self.oilValueObservable = oilPercentageObservable.map { "\(Int($0))%"}.startWith("0%")
```

**Comment by dominicfreeston** [comment]
Observable as input and output to a system really appeals to me. When I started with RxSwift, this article got me thinking that way: http://staltz.com/unidirectional-user-interface-architectures.html.

One of the main reasons we didn't go with this approach is the risk of retain cycles, where your service and your view end up retaining each other via the subscriptions. I'm really interested to see what you come up with this, but I'd recommend you make sure things don't live beyond their expected lifespan (which is thankfully really easy with a deterministic memory management system like ARC).

File: VespaGASTests/ViewModels/GasCalculatorViewModelTests.swift (hunk 2 above)
Comment (amlcurran) [inline VespaGASTests/ViewModels/GasCalculatorViewModelTests.swift]: RxBlocking is highly recommended for tests, I'd argue it should be required for all new projects (especially as the blocking stuff is in RxJava main lib)

File: VespaGAS/Managers/OilMixCalculator.swift [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+typealias GasType = Float
+typealias OilType = Float
+typealias OilMixType = Float
+
+let MLInLiter: Float = 1000
+
+struct OilMixCalculator {
+    static func oilMLForGasoline(gasInLiter: GasType, oilPercentage: OilMixType) -> OilType? {
+        if oilPercentage == 0 {
+            return nil
```
Comment (amlcurran) [inline VespaGAS/Managers/OilMixCalculator.swift]: In terms of Domain Design, I think this might make better sense that this returns a non-Optional `OilType` and this branch returns `0`. If you think about it, you don't add a `nil` amount of oil to an engine, you'd add 0ml :) 

this may be a good book if you're interested in domain driven design - it is very short and I highly recommend reading it https://www.infoq.com/minibooks/domain-driven-design-quickly

File: VespaGAS/Managers/OilMixCalculator.swift [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+        }
+
+        return gasInLiter * MLInLiter / 100 * oilPercentage
```
Comment (amlcurran) [inline VespaGAS/Managers/OilMixCalculator.swift]: furthermore, whilst you've used typealiases this is still a classic case of connascence of meaning :) you must know that `GasType` is measure in litres and `OilType` is returned in litres in order for this to be correct. 

I'd recommend creating a `Volume` struct which encapsulates this better. Furthermore, this struct could have behaviour on it. Typically if you have a static function acting on a value object, then that could be a method on the value object itself. 

e.g. compare

`Calculator.multiply(someDomainObject, 4)`

to

`someDomainObject.multiplyBy(4)`

File: VespaGAS/ViewModels/GasCalculatorViewModel.swift [hunk 5]
```diff
@@ -1,5 +1,36 @@
...
+    var oilValueObservable: Observable<String> {
+        return oilValueSubject.asObservable()
+    }
+    var gasValueObservable: Observable<String> {
+        return gasValueSubject.asObservable()
+    }
+
+    init(gasObservable: Observable<GasType>, oilPercentageObservable: Observable<OilMixType>) {
+        oilMixObservable = Observable.combineLatest(gasObservable, oilPercentageObservable) { (gasInLiter, oilPercentage) -> String in
+            if let oilValue = OilMixCalculator.oilMLForGasoline(gasInLiter, oilPercentage: oilPercentage) {
```
Comment (amlcurran) [inline VespaGAS/ViewModels/GasCalculatorViewModel.swift]: this is some logic that is unrelated to the view and to be doesn't make sense to be in a view model. This is actually the _model_ in my opinion. Might be worth splitting it out?

**Comment by gbasile** [comment]
I have updated the PR with the changes proposed by @amlcurran 

Now I have a dedicated struct for the Volume while Gasoline and Oil are two typealias of this new type. 
The reason I'm using a typealias is because I can't inherit from a struct, I'm not sure if this one is a good idea. 

OilMix is a new type as well, and it represents the percentage of the oil Mixture. 

I have to say that using types linked to the domain has made the code much more readable and expressive. I don't want to say that it can be read as "well written prose" (cit. Grady Booch)  but there is a good improvement on the readability.

File: VespaGAS/Models/Volume.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+enum VolumeUnit: CustomStringConvertible {
```
Comment (amlcurran) [inline VespaGAS/Models/Volume.swift:3]: this is an interesting way of doing this! 👍

File: VespaGAS/Managers/OilMixCalculator.swift [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+let MLInLiter: Float = 1000
+
+typealias Oil = Volume
+typealias Gasoline = Volume
+
+struct OilMixCalculator {
+    static func oilFor(gasoline: Gasoline, oilMix: OilMix) -> Oil {
```
Comment (amlcurran) [inline VespaGAS/Managers/OilMixCalculator.swift]: you still have a static method here, how about moving this further into your domain?  

in TDD it is very easy to end up with what you have here which is called the [Anaemic domain model](http://www.martinfowler.com/bliki/AnemicDomainModel.html), where your domain objects are just bunches of data and you end up with static functions like this which do all the work. It is really worth trying to get around this because it makes the code far more approachable to someone coming on to a project and therefore reduces maintainence overhead 😸 it also means your code matches the domain more closely.

If you think of a concrete example - a new Java developer won't know how to square root a number. They'd probably expect something like `1.sqrt()` but instead they have to Google and find that it is instead `Math.sqrt(1)`!

**Comment by amlcurran** [comment]
Just one comment to discuss but looks great otherwise 💯

File: VespaGAS/Managers/OilMixCalculator.swift (hunk 7 above)
Comment (gbasile) [inline VespaGAS/Managers/OilMixCalculator.swift]: You are absolutely right! 

In my previous iteration I have created this calculator because it was not clear who was responsible of this logic and there were no models for the Gasoline and the Oil. 
With the introduction of the Oil model it would make sense to have this method on this object, so I could introduce something like: 

`static func oilFor(gasoline: Gasoline, oilMix: OilMix)`
or 
`init(gasoline: Gasoline, oilMix: OilMix)`

What do you think?

File: VespaGAS/Models/Volume.swift (hunk 6 above)
Comment (gbasile) [inline VespaGAS/Models/Volume.swift:3]: Hope _interesting_ is not the english polite way to say _weird_ 😉

Check #9 of this chart: 
language-web

Comment (amlcurran) [inline VespaGAS/Models/Volume.swift:3]: haha, it doesn't this time no :) I was trying to think I how I would do this but I'm not sure, it clearly works well!

File: VespaGAS/Managers/OilMixCalculator.swift (hunk 7 above)
Comment (amlcurran) [inline VespaGAS/Managers/OilMixCalculator.swift]: Cool! Yeah having domain objects will make it easier to see things like this. Domain-driven design is really powerful and you'll have noticed we used it significantly in OC.

You could add this behaviour to `Oil` or `Gasoline` but that would mean altering the underlying `Volume` type right? Are there any other domain objects this behaviour could make sense on?

File: VespaGAS/ViewModels/GasCalculatorViewModel.swift [hunk 8]
```diff
@@ -1,5 +1,38 @@
...
+    init(gasObservable: Observable<Gasoline>, oilMixObservable: Observable<OilMix>) {
+        oilMixValueObservable = Observable.combineLatest(gasObservable, oilMixObservable) { (gasoline, oilMix) -> String in
+            let oil = oilMix.oilForGasoline(gasoline)
+            return oil.toUnit(.Milliliter).roundAmount().description.lowercaseString
+        }.startWith(Oil(amount: 0, unit: .Milliliter).description.lowercaseString)
+
+        oilMixObservable.subscribeNext() { oilMix in
+            let oilSubject = self.oilValueSubject
+            oilSubject.onNext(oilMix)
+        }.addDisposableTo(disposeBag)
```
Comment (gbasile) [inline VespaGAS/ViewModels/GasCalculatorViewModel.swift]: Here to split this in 2 lines is intentional, If I use: 

`self.oilValueSubject.onNext(oilMix)` 

I have a segmentation fault during compilation 😐

The same for the next block
Any idea?

File: VespaGAS/Managers/OilMixCalculator.swift (hunk 7 above)
Comment (gbasile) [inline VespaGAS/Managers/OilMixCalculator.swift]: I have implemented this logic now into the OilMix object. Changes are already here.

File: VespaGAS/ViewModels/GasCalculatorViewModel.swift (hunk 8 above)
Comment (amlcurran) [inline VespaGAS/ViewModels/GasCalculatorViewModel.swift]: no, I know that has been seen before, I think @dominicfreeston might have seen this?

File: VespaGAS.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -8,6 +8,7 @@
```
Comment (gbasile) [inline VespaGAS.xcodeproj/project.pbxproj:8]: I believe this commit is interesting and I would like to discuss why.

_Oil_ and _Gasoline_ were not strong typed: they were just typealias of the _Volume_ type. 
For this simple reason, I could have changed every instance of _Oil_ for _Gasoline_ without any 
complain from the compiler; a first thought was to have _Oil_ and _Gasoline_ to inherit from a base
class _Volume_, but this would have introduced reference types while I wanted to work with value 
types. A better and swiftly solution has been to have a _Volume_ protocol, a protocol extensions 
for the default implementations, and make the structs _Oil_ and _Gasoline_ to conform this protocol. 

While writing the protocol extensions I have noticed that several methods were returning new modified copy
of the object, and for this reason I needed to use a Generic Protocol, not just a Protocol. 

A generic protocol is not a type as a protocol or a struct, it's less type than any other type; in fact we can only use it 
as a generic constraints. In this case I haven't to face all the limitations of a Generic Protocol, instead I  only had to use the `Self` associated type when required in the methods and some little modification of the equality function. 

I was not able to reuse the equality function from _Volume_: 

```
func ==(lhs: Volume, rhs: Volume) -> Bool {
    return lhs.amount == rhs.amount && lhs.unit == rhs.unit
}

```

The compiler would have complained that now _Volume_ has an associated type and that I could only use it as a generic constraints, the solution is to simply use it as a constraints. This concept is so new in Swift that the first time I had to think a little about it in order to understand it. To use _Volume_ as a generic constraints is that easy as to: 

```
func ==<A where A:Volume>(lhs: A, rhs: A) -> Bool {
    return lhs.amount == rhs.amount && lhs.unit == rhs.unit   
}
```

Now I have most of the behaviour of _Oil_ and _Gasoline_ in the protocol extension of _Volume_, and the code I need to write to implement the new _Oil_ and _Gasoline_ is just few lines:
`struct Oil: Volume {
    var amount: Float
    var unit: VolumeUnit
}
`

Unfortunately I can't declare amount and unit in the protocol extension, so I have to do it into each of the types that implement the _Volume_ protocol 😞 I don't know if there is a solution to avoid these 2 lines in each type, please tell me if you have some idea because these 2 lines are there basically to make the compiler happy.

PS: The behaviours of the protocol extension have been tested using a simple MockVolume struct in the test target. Only _Gasoline_ has one test because of the different behaviour of the CustomStringConvertible protocol. 

Wow, I have written a lot! Time to go sleep!

File: VespaGAS/Models/Volume.swift [hunk 10]
```diff
@@ -23,26 +23,34 @@ enum VolumeUnit: CustomStringConvertible {
...
 
-    func roundAmount() -> Volume {
-        return Volume(amount: round(self.amount), unit: unit)
+    func roundAmount() -> Self {
+        return Self(amount: round(amount), unit: unit)
     }
 }
 
-func ==(lhs: Volume, rhs: Volume) -> Bool {
+func ==<A where A:Volume>(lhs: A, rhs: A) -> Bool {
```
Comment (amlcurran) [inline VespaGAS/Models/Volume.swift:54]: so with this, would an `Oil` with a certain volume and  a `Petrol` with the same volume be considered equal?

File: VespaGAS/Models/Gasoline.swift [hunk 11]
```diff
@@ -1 +1,11 @@
-typealias Gasoline = Volume
+import Foundation
+
+struct Gasoline: Volume {
+    var amount: Float
```
Comment (amlcurran) [inline VespaGAS/Models/Gasoline.swift:4]: I believe these could be `let` - think it only has to be `var` in the protocol declaration

File: VespaGAS/Models/Volume.swift (hunk 10 above)
Comment (gbasile) [inline VespaGAS/Models/Volume.swift:54]: Nope, you can only compare two objects implementing the same type. This makes sense for me because Oil and Gasoline are not the same even if they have the same amount and unit. 
If you want to compare any type implementing the protocol you can say something like: 

`func ==<A, B where A:Volume, B: Volume>(lhs: A, rhs: B) -> Bool`

File: VespaGAS/Models/Gasoline.swift (hunk 11 above)
Comment (gbasile) [inline VespaGAS/Models/Gasoline.swift:4]: Good spot! I'm going to change it even if I would like to remove them. 
As I said in the comment at the beginning of the commit, I would like to remove these 2 lines as they are only there to make the compiler happy. Do you have any idea?

Comment (amlcurran) [inline VespaGAS/Models/Gasoline.swift:4]: I think you must define them unfortunately.

File: VespaGAS/Models/Volume.swift (hunk 10 above)
Comment (amlcurran) [inline VespaGAS/Models/Volume.swift:54]: ah of course! Silly me :) did you see the syntax for `where` statements are changing in Swift 3?  https://github.com/apple/swift/commit/c5bf433490f4fe471584550b431695498a6adc0e

Comment (gbasile) [inline VespaGAS/Models/Volume.swift:54]: wow, I don't know if I like the new syntax. Probably it makes sense when you have a lot of where conditions.

File: VespaGAS/Models/Gasoline.swift (hunk 11 above)
Comment (gbasile) [inline VespaGAS/Models/Gasoline.swift:4]: Apparently I can't use let. Compiler complains about these structs not conforming anymore to the Volume protocol

File: VespaGAS/ViewModels/GasCalculatorViewModel.swift (hunk 1 above)
Comment (gbasile) [inline VespaGAS/ViewModels/GasCalculatorViewModel.swift]: Good idea, I have refactored the ViewModel with your proposal and now it looks way better! 👍 
Just a consideration about memory management, am I doing it right? Do I need to dispose something in this ViewModel?

File: VespaGAS.xcodeproj/project.pbxproj (hunk 9 above)
Comment (dominicfreeston) [inline VespaGAS.xcodeproj/project.pbxproj:8]: I have a bunch of vague thoughts I might come back to later, but my first question is about conforming to `Equatable`. I'm assuming neither `Volume`, nor `Gasoline` nor `Oil` conform to it (having an `==` function is not the same as conforming to `Equatable`). Maybe a little code repetition is actually better in this case (i.e. implement `==` for each of them)?

Comment (gbasile) [inline VespaGAS.xcodeproj/project.pbxproj:8]: What do you mean by "having a `==` function is not the same as conforming to Equatable"? 

The provided `==` function cover both Oil and Gasoline, why should I provided another custom implementation of `==` if they are already covered by the generic one?

File: VespaGASTests/ViewModels/GasCalculatorViewModelTests.swift (hunk 2 above)
Comment (gbasile) [inline VespaGASTests/ViewModels/GasCalculatorViewModelTests.swift]: I have updated the PR using RxTest. 
I like the way the tests looks now 👍

**Comment by gbasile** [comment]
For me the PR is ready to be merged. I Need your approval or your new comments

File: VespaGAS.xcodeproj/project.pbxproj (hunk 9 above)
Comment (wltrup) [inline VespaGAS.xcodeproj/project.pbxproj:8]: I think what @dominicfreeston means is that just providing the `==` function isn't sufficient; you need to explicitly say that a type conforms to `Equatable`, i.e., you also need to add `: Equatable` to the type declaration.

Comment (wltrup) [inline VespaGAS.xcodeproj/project.pbxproj:8]: " I can't declare amount and unit in the protocol extension"

True and unfortunate. There is a solution, but I don't like it, you're not going to like it, and I'm sure @dominicfreeston doesn't like it. As you probably know, all `NSObject` instances have a dictionary property that you can add stuff to so you could add properties to an object at runtime, even from within a protocol. However, that means you'd have to give up the value nature of your types and make them classes (not just any classes either - they'd have to subclass `NSObject`). If you're sticking to value types, then I'm not aware of any way to get around the fact that you can't add stored property declarations to protocol extensions.

**Comment by wltrup** [comment]
It looks good to me but, in the future, it would be nice to have smaller commits/PRs. You went from nearly no implementation (just barely out of an Xcode template) to a fairly complete implementation, all in one commit/PR !

File: VespaGAS.xcodeproj/project.pbxproj (hunk 9 above)
Comment (gbasile) [inline VespaGAS.xcodeproj/project.pbxproj:8]: >  you need to explicitly say that a type conforms to Equatable, i.e., you also need to add : Equatable to the type declaration.

Both Gasoline and Oil conforms to Equatable because Volume does it in first instance. 
I don't see the value in specifying again for Gasoline and Oil that they are equatable. I could provide a custom implementation for both of them, but the result will be the same. 

We want to be sure that: 
1. 2 Volume objects are equal if they have the same amount and unit
2. 2 Gasoline objects are equal if they have the same amount and unit
3. 2 Oil objects are equals if they have the same amount and unit
4. Anything else is not equal: I can't even compare Oil with Gasoline with this implementation

Comment (wltrup) [inline VespaGAS.xcodeproj/project.pbxproj:8]: No, if `Volume` already does then it's ok. I was just elaborating on Dom's response.

I think your list of equality conditions just above is the correct way to implement equality.

**Comment by gbasile** [comment]
@dominicfreeston @amlcurran @wltrup @yvettecook @eduardourso thumbs up?

**Comment by wltrup** [comment]
👍

**Comment by amlcurran** [comment]
sure, but we don't have merge permission ;)

File: VespaGAS/Models/Gasoline.swift [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+    var unit: VolumeUnit
+
+    var description: String {
```
Comment (amlcurran) [inline VespaGAS/Models/Gasoline.swift:7]: why is this declared even though it is declared in the protocol extension?

Comment (wltrup) [inline VespaGAS/Models/Gasoline.swift:7]: Because it's overriding the implementation provided by the protocol extension. Compare the implementation in the protocol extension to `Volume` with this one and you'll see that they're different.

**Comment by amlcurran** [comment]
Final comment - I feel like there is still some way of improving the gasoline/oil/volume part, although I'm not sure how. I'd generally not expect a value object (volume) to be an interface, because it represents a concrete thing (the _small_ code smell here is having properties on the protocol). 

The other option would to have Gasoline and Oil have a _property_ of volume ("I have gasoline, it has a volume of X" as opposed to "I have gasoline which is a volume of X"). But I think the design is fine right now and am not convinced that the suggestion I just made would be better. Just might be something to consider if you ever revisited the code :)

File: VespaGAS/Models/Gasoline.swift (hunk 12 above)
Comment (gbasile) [inline VespaGAS/Models/Gasoline.swift:7]: @wltrup that's correct. Having a look at the Unit tests makes it more clear.
